brayns/io/MorphologyLoader.cpp

```diff
@@ -313,18 +328,290 @@ class MorphologyLoader::Impl
+                const float radiusSum = geom.radius + bifGeom.radius;
+
+                if (dist0 < radiusSum || dist1 < radiusSum)
+                {
+                    sdfMorphologyData.neighbours[bifId].insert(geomIdx);
+                    sdfMorphologyData.neighbours[geomIdx].insert(bifId);
+                }
+            }
+        }
+    }
+
+    /**
+     * @brief _finalizeSDFGeometries Calculates all neighbours and adds the
+     * geometries to the model.
      */
+    void _finalizeSDFGeometries(Model& model,
+                                SDFMorphologyData& sdfMorphologyData) const
+    {
+        const size_t numGeoms = sdfMorphologyData.geometries.size();
+        sdfMorphologyData.localToGlobalIdx.resize(numGeoms, 0);
+
+        // Extend neighbours to make sure smoothing is applied on all
+        // closely connected geometries
+        for (size_t rep = 0; rep < 4; rep++)
+        {
+            const size_t numNeighs = sdfMorphologyData.neighbours.size();
+            auto neighsCopy = sdfMorphologyData.neighbours;
+            for (size_t i = 0; i < numNeighs; i++)
+            {
+                for (size_t j : sdfMorphologyData.neighbours[i])
+                {
+                    for (size_t newNei : sdfMorphologyData.neighbours[j])
+                    {
+                        neighsCopy[i].insert(newNei);
+                        neighsCopy[newNei].insert(i);
+                    }
+                }
+            }
+            sdfMorphologyData.neighbours = neighsCopy;
+        }
+
+        for (size_t i = 0; i < numGeoms; i++)
+        {
+            sdfMorphologyData.localToGlobalIdx[i] =
+                model.addSDFGeometry(sdfMorphologyData.materials[i],
+                                     sdfMorphologyData.geometries[i], {});
+        }
+
+        for (size_t i = 0; i < numGeoms; i++)
+        {
+            // Convert neighbours from set to vector and erase itself from its
+            // neighbours
+            std::vector<size_t> neighbours;
+            const auto& neighSet = sdfMorphologyData.neighbours[i];
+            std::copy(neighSet.begin(), neighSet.end(),
+                      std::back_inserter(neighbours));
+            neighbours.erase(
+                std::remove_if(neighbours.begin(), neighbours.end(),
+                               [i](size_t elem) { return elem == i; }),
+                neighbours.end());
+
+            for (auto& idx : neighbours)
+                idx = sdfMorphologyData.localToGlobalIdx[idx];
+
+            model.setSDFGeometryNeighbours(
+                sdfMorphologyData.localToGlobalIdx[i], neighbours);
+        }
+    }
+
+    /**
+     * @brief _calculateMorphologyTreeStructure Calculates the structure of the
+     * morphology tree by finding overlapping beginnings and endings of the
+     * sections.
+     */
+    MorphologyTreeStructure _calculateMorphologyTreeStructure(
+        const brain::neuron::Sections& sections) const
```

Comment:
this function is very long and could be refactored into smaller parts

Reply:
I left this as it is but added some comments since I think the method is actually quite straightforward in its current state. If you disagree I can fix it.